README.md
# hcipher

hcipher api encode


CHANGES.md
Change Log

## v1.3.0
    [ Release Date: 2021-04-05 ]
- [feature] v3 AES+SHA1+RSA

## v1.0.0
    [ Release Date: 2021-04-05 ]
- [feature] v1 DES+MD5
- [feature] v2 AES+RSA

v1/desmd5_test.go
package v1

import (
	"net/url"
	"testing"
)

func TestEncryptDesMd5(t *testing.T) {
	key := "88888888"
	s := `{"appid":"88888888"}`
	/*
		密文 e2a19818214e2c1523208747dd872fa9xmF4oTvCx2mrAp8erL7oHg==
		明文 123中文
	*/

	//加密
	v, e := EnDesMd5(s, key)
	if e != nil {
		t.Error(e.Error())
		return
	}
	t.Log(v)
	t.Log(url.QueryEscape(v))

	// 解密
	v, e = DeDesMd5(v, key)
	if e != nil {
		t.Error(e.Error())
		return
	}
	t.Log(v)
}


v3/aes_sha1_rsa_test.go
package v3

import "testing"

func TestEnCode(t *testing.T) {
	pubKey := "MIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQCf0arfpOm8uRJ/HUg3HUzGnvXF9L18tlFvEAas59XZ5GAxQSIVyK3+yAxo4YQ1zjcHCFDk0zvEHvb7MPNSulu6zxDoyskCLVhTfzmWncewyAtu1YZGpIZaI496JtBsDjb2uszR8FrPhDoC4tyiEDVGMyGDa2hruQfXExrMSymY/wIDAQAB"
	plaintext := "{123}"

	ciphertext, err := Encode(plaintext, pubKey)
	if err != nil {
		t.Error(err)
		return
	}
	t.Log("加密结果：" + ciphertext)

	priKey := "MIICdgIBADANBgkqhkiG9w0BAQEFAASCAmAwggJcAgEAAoGBAJ/Rqt+k6by5En8dSDcdTMae9cX0vXy2UW8QBqzn1dnkYDFBIhXIrf7IDGjhhDXONwcIUOTTO8Qe9vsw81K6W7rPEOjKyQItWFN/OZadx7DIC27Vhkakhlojj3om0GwONva6zNHwWs+EOgLi3KIQNUYzIYNraGu5B9cTGsxLKZj/AgMBAAECgYBYyOnchIk1Rrt30pSjyas1TGnNJ0F9Xuuuf4C13bV37t9hfYWqBGpk+E2sllwUaBM53OzTsmWpdmpO8cq7UrxSEqkOJU4zUtdD7031eTgq9Fyp53EBU54mKOAOVoBoHdu/Qfv6GI8X2vmlf6FFeqheWST/AZ/XPkbG5Pp3zAa5QQJBAMpm7usIR3Vdqyl1OVBm4IU+eFTdDDPl31T/CR7jaUAH+d9KFedJOM2JdyQPwKFsRTwEEwBm9vgNrk3Kq6PAEQsCQQDKI/rJG/b04ZP7af28qPRmEquozdevsDe0s33dzWhLimqeriHDI8+uhcqgGIC1gh62LWnYpdEE8VkCDg4LmzhdAkEArMTskEeC59ZK8pqTj+QhJtvKT3ZYojxIRP9mQ62O/A9S5Z0R0VmZWSlMP8YKgkAvYSmBJsi9a8QR02l61c5vPQJAbr3HZuYrJX1v1Qz8NZ9aRZF0+cXLpDSmUBkFm74spTXvs38yf/XekX46w/qoiMgAi03V7xroqAyQ9s88Yp9nAQJAXB3jqkPjSpeDzn09JqjdKJ5XzqSidA4/vhZotbIC24rDBN1+dcB67Lx2ggicZYrjpc70CjBgoBR+I6w7qpatQg=="
	r, err := Decode(ciphertext, priKey)
	if err != nil {
		t.Error(err)
		return
	}
	t.Log("解密结果：" + r)
}


v1/desmd5.go
package v1

import (
	"bytes"
	"crypto/cipher"
	"crypto/des"
	"crypto/md5"
	"encoding/base64"
	"errors"
	"fmt"
)

/*
对称加密（DES）+验签（MD5）方式方法如下：

加密步骤：
1. 【加密报文】通过双方约定的密钥K1使用DES加密报文明文S1，得到密文字节数组B1;
2. 【加密报文】取得密文字节数组的Base64表示形式S2;
3. 【计算签名】将双方约定的密钥K1拼接在S2的前部，得到S3;
4. 【计算签名】计算S3的MD5，并取得Hex字符串的小写形式M1;
5. 【拼接结果】将M1和S2拼接得到最终密文报文：R

解密步骤：
1. 【拆解密文】取得密文报文字符串S1的前32个字符，得到签名M1，取得32个字符后的内容，得到S2;
2. 【验证签名】将双方约定的密钥K1拼接在S2的前部，得到S3;
3. 【验证签名】计算S3的MD5，并取得Hex字符串的小写形式M2;
4. 【验证签名】判断M1是否等于M2，不相等则验签失败，否则取得S2的反Base64后的密文字节数组B1;
5. 【解密报文】通过双方约定的密钥K1使用DES解密密文字节数组B1，得到明文报文：R

*/

// DES+MD5接口报文加密
func EnDesMd5(s, key string) (v string, err error) {
	k1 := []byte(key)
	s1 := []byte(s)

	block, err := des.NewCipher(k1)
	if err != nil {
		return
	}
	// PKCS5Padding
	blockSize := block.BlockSize()
	padding := blockSize - len(s1)%blockSize
	paddingText := bytes.Repeat([]byte{byte(padding)}, padding)
	s1 = append(s1, paddingText...)

	//DES CBCEncrypt
	blockMode := cipher.NewCBCEncrypter(block, k1)
	cipherText := make([]byte, len(s1))
	blockMode.CryptBlocks(cipherText, s1)

	// Base64 Encode
	s2 := base64.StdEncoding.EncodeToString(cipherText)

	// Md5 签名
	s3 := key + s2
	m1 := fmt.Sprintf("%x", md5.Sum([]byte(s3)))
	r := m1 + s2
	return r, err
}

// DES+MD5接口报文解密
func DeDesMd5(s, key string) (v string, err error) {
	if len(s) < 32 {
		err = errors.New("Error:Ciphertext format error")
		return
	}
	m1 := s[0:32]
	s2 := s[32:]

	// MD5签名验证
	s3 := key + s2
	m2 := fmt.Sprintf("%x", md5.Sum([]byte(s3)))
	if m1 != m2 {
		err = errors.New("Error:Signature verification failed")
		return
	}

	// Base64 Decode
	b1, err := base64.StdEncoding.DecodeString(s2)
	if err != nil {
		return
	}

	k1 := []byte(key)
	block, err := des.NewCipher(k1)
	if err != nil {
		return
	}

	//DES CBCDecrypt
	blockMode := cipher.NewCBCDecrypter(block, k1)
	text := make([]byte, len(b1))
	blockMode.CryptBlocks(text, b1)

	// PKCS5UnPadding
	length := len(text)
	unPadding := int(text[length-1])
	text = text[:(length - unPadding)]

	r := string(text)
	return r, err
}


v2/aes_rsa.go
package v2

import (
	"bytes"
	"crypto/aes"
	"crypto/cipher"
	"crypto/md5"
	"crypto/rand"
	"crypto/rsa"
	"crypto/x509"
	"encoding/base64"
	"encoding/pem"
	"errors"
	"fmt"
	"strings"
)

/**
报文加密（AES）+密钥加密（RSA）方式方法如下：

加密步骤：
1. 【生成秘钥】生成随机32位字符串s1作为AES秘钥；
2. 【加密密钥】使用RSA方式加密AES密钥输出base64格式字符串s2
3. 【加密报文】AES加密报文输出base64格式字符串s3
4. 【加密结果】使用冒号拼接结果 ciphertext = s2 + ":" + s3 ；

解密步骤：
1. 【拆解密文】根据冒号拆分密文并做baes64解码，密文报文s3，密文AES密钥s2；
2. 【解密密钥】使用RSA解密密钥s2获得AES密钥s1；
3. 【解密报文】通过密钥s1使用AES解密密文s3；
4. 【解密结果】获得s3明文plaintext

注意：公钥私钥都输出为base64格式字符串保存，非pem格式。
AES CBC 256
RSA PKCS8 1024
*/

// AES+RSA 加密报文
func EnAesRsa(plaintext, pubKey string) (ciphertext string, err error) {
	defer func() {
		if e := recover(); e != nil {
			err = e.(error)
		}
		if err != nil {
			err = errors.New("EN_AES_RSA: 加密出错 " + err.Error())
		}
	}()

	// 随机32位字符串作为AES密钥
	s1 := Rand32()
	aesKey := []byte(s1)

	// RSA  公钥支持pem格式与base格式
	var derBytes []byte
	if strings.Contains(pubKey, "BEGIN PUBLIC KEY") {
		p, _ := pem.Decode([]byte(pubKey))
		if p == nil {
			err = errors.New("pem.Decode error ")
			return
		}
		derBytes = p.Bytes
	} else {
		derBytes, err = base64.StdEncoding.DecodeString(pubKey)
		if err != nil {
			err = errors.New("base64.Decode error " + err.Error())
			return
		}
	}

	pub, err := x509.ParsePKIXPublicKey(derBytes)
	if err != nil {
		err = errors.New("x509.ParsePKIXPublicKey error " + err.Error())
		return
	}
	aes_key_ciphertext, err := rsa.EncryptPKCS1v15(rand.Reader, pub.(*rsa.PublicKey), aesKey)
	if err != nil {
		err = errors.New("rsa.EncryptPKCS1v15 error " + err.Error())
		return
	}
	s2 := base64.StdEncoding.EncodeToString(aes_key_ciphertext)

	// AES CBC 加密
	block, err := aes.NewCipher(aesKey)
	if err != nil {
		err = errors.New("aes.NewCipher error " + err.Error())
		return
	}
	plainText := Padding([]byte(plaintext), block.BlockSize())
	blockMode := cipher.NewCBCEncrypter(block, []byte(s1[0:block.BlockSize()]))
	cipherText := make([]byte, len(plainText))
	blockMode.CryptBlocks(cipherText, plainText)
	s3 := base64.StdEncoding.EncodeToString(cipherText)

	// 密文结果
	ciphertext = s2 + ":" + s3
	return
}

// AES+RSA 解密报文
func DeAesRsa(ciphertext, priKey string) (plaintext string, err error) {
	defer func() {
		if e := recover(); e != nil {
			err = e.(error)
		}
		if err != nil {
			err = errors.New("DE_AES_RSA:解密出错" + err.Error())
		}
	}()
	arr := strings.Split(ciphertext, ":")
	s2 := arr[0]
	s3 := arr[1]

	aes_key_ciphertext, err := base64.StdEncoding.DecodeString(s2)
	if err != nil {
		err = errors.New("base64.Decode aesKey error" + err.Error())
		return
	}
	cipherText, err := base64.StdEncoding.DecodeString(s3)
	if err != nil {
		err = errors.New("base64.Decode ciphertext  error" + err.Error())
		return
	}

	// RSA  公钥支持pem格式与base格式
	var der []byte
	if strings.Contains(priKey, "BEGIN PRIVATE KEY") {
		p, _ := pem.Decode([]byte(priKey))
		if p == nil {
			err = errors.New("pem.Decode privateKey error ")
			return
		}
		der = p.Bytes
	} else {
		der, err = base64.StdEncoding.DecodeString(priKey)
		if err != nil {
			err = errors.New("base64.Decode privateKey error" + err.Error())
			return
		}
	}

	prikey, err := x509.ParsePKCS8PrivateKey(der)
	if err != nil {
		err = errors.New("x509.ParsePKCS8PrivateKey error " + err.Error())
		return
	}
	key, err := rsa.DecryptPKCS1v15(rand.Reader, prikey.(*rsa.PrivateKey), aes_key_ciphertext)
	if err != nil {
		err = errors.New("rsa.DecryptPKCS1v15 error " + err.Error())
		return
	}

	// AES 解密
	block, err := aes.NewCipher(key)
	if err != nil {
		err = errors.New("aes.NewCipher error" + err.Error())
		return
	}
	blockMode := cipher.NewCBCDecrypter(block, []byte(string(key)[0:block.BlockSize()]))
	plainText := make([]byte, len(cipherText))
	blockMode.CryptBlocks(plainText, cipherText)
	plainText = UnPadding(plainText)

	// 明文结果
	plaintext = string(plainText)
	return
}

func Padding(plainText []byte, blockSize int) []byte {
	n := blockSize - len(plainText)%blockSize //计算要填充的长度
	temp := bytes.Repeat([]byte{byte(n)}, n)  //对原来的明文填充n个n
	plainText = append(plainText, temp...)
	return plainText
}

func UnPadding(cipherText []byte) []byte {
	end := cipherText[len(cipherText)-1]               //取出密文最后一个字节end
	cipherText = cipherText[:len(cipherText)-int(end)] //删除填充
	return cipherText
}

// Rand32 生成32位随机字符串 使用crypto/rand 随机赋值byte数组， 然后md5返回32位十六进制字符串
func Rand32() string {
	var b = make([]byte, 48)
	if _, err := rand.Read(b); err != nil {
		return ""
	}
	return fmt.Sprintf("%x", md5.Sum(b))
}


v3/aes_sha1_rsa.go
package v3

import (
	"bytes"
	"crypto/aes"
	"crypto/cipher"
	"crypto/md5"
	"crypto/rand"
	"crypto/rsa"
	"crypto/sha1"
	"crypto/x509"
	"encoding/base64"
	"encoding/pem"
	"errors"
	"fmt"
	"strings"
)

/**
V3 报文加密（AES）+ 签名（SHA1） +密钥加密（RSA）方式方法如下：

加密步骤：
1. 明文字符串 plaintext ;
2. 生成16位随机字符串 ak  作为AES与SHA1的秘钥 ;
3. AES-CBC加密 plaintext 输出base64字符串 s1 ;
4. SHA1计算字符串 s1+ak 签名，输出 s2 ;
5. RSA加密 ak 输出s3 ;
6. 获取最终密文 ciphertext = s1:s2:s3

解密步骤：
1. 密文字符串 ciphertext
2. 根据冒号拆分获取s1,s2,s3;
3. RSA 解密s3，获取 ak ，用于AES解密与SHA1验签；
4. SHA1 计算 s1+ak 签名 sign，
5. 验证签名sign是否等于s2；
6. AES解密s1获取明文 plaintext ;

注意：公钥私钥都输出为base64格式字符串保存，非pem格式。
AES CBC 128
RSA PKCS8 1024
*/

// Encode AES+SHA1+RSA 加密报文
func Encode(plaintext, pubKey string) (ciphertext string, err error) {
	defer func() {
		if e := recover(); e != nil {
			err = e.(error)
		}
		if err != nil {
			err = errors.New("ENCODE_V3_ERROR: 加密出错 " + err.Error())
		}
	}()

	// 随机32位字符串作为AES密钥
	aKey := rand32()[8:24]
	aesKey := []byte(aKey)

	// AES CBC 128 加密
	block, err := aes.NewCipher(aesKey)
	if err != nil {
		err = errors.New("aes.NewCipher error " + err.Error())
		return
	}
	plainText := Padding([]byte(plaintext), block.BlockSize())
	blockMode := cipher.NewCBCEncrypter(block, []byte(aKey[0:block.BlockSize()]))
	cipherText := make([]byte, len(plainText))
	blockMode.CryptBlocks(cipherText, plainText)
	s1 := base64.StdEncoding.EncodeToString(cipherText)

	// SHA1签名
	s2 := fmt.Sprintf("%x", sha1.Sum([]byte(s1+aKey)))

	// RSA  公钥支持pem格式与base格式
	var derBytes []byte
	if strings.Contains(pubKey, "BEGIN PUBLIC KEY") {
		p, _ := pem.Decode([]byte(pubKey))
		if p == nil {
			err = errors.New("pubKey pem.Decode error ")
			return
		}
		derBytes = p.Bytes
	} else {
		derBytes, err = base64.StdEncoding.DecodeString(pubKey)
		if err != nil {
			err = errors.New("pubKey Base64解码错误 " + err.Error())
			return
		}
	}

	pub, err := x509.ParsePKIXPublicKey(derBytes)
	if err != nil {
		err = errors.New("x509.ParsePKIXPublicKey error " + err.Error())
		return
	}
	aesKeyCiphertext, err := rsa.EncryptPKCS1v15(rand.Reader, pub.(*rsa.PublicKey), aesKey)
	if err != nil {
		err = errors.New("rsa.EncryptPKCS1v15 error " + err.Error())
		return
	}
	s3 := base64.StdEncoding.EncodeToString(aesKeyCiphertext)

	// 密文结果 =  报文密文:签名:密钥密文
	ciphertext = s1 + ":" + s2 + ":" + s3
	return
}

// Decode AES+SHA1+RSA 解密报文
func Decode(ciphertext, priKey string) (plaintext string, err error) {
	defer func() {
		if e := recover(); e != nil {
			err = e.(error)
		}
		if err != nil {
			err = errors.New("DECODE_V3_ERROR:" + err.Error())
		}
	}()
	split := strings.Split(ciphertext, ":")
	s1 := split[0] // 密文报文
	s2 := split[1] // s1的签名，需SHA1验签
	s3 := split[2] // s1的秘钥，需RSA解密

	// RSA  公钥支持pem格式与base格式
	var der []byte
	if strings.Contains(priKey, "BEGIN PRIVATE KEY") {
		p, _ := pem.Decode([]byte(priKey))
		if p == nil {
			err = errors.New("priKey pem.Decode error ")
			return
		}
		der = p.Bytes
	} else {
		der, err = base64.StdEncoding.DecodeString(priKey)
		if err != nil {
			err = errors.New("priKey Base64解码错误 " + err.Error())
			return
		}
	}

	// RSA 解密
	aesKeyCiphertext, err := base64.StdEncoding.DecodeString(s3)
	if err != nil {
		err = errors.New("s3 Base64解码错误 " + err.Error())
		return
	}
	privateKey, err := x509.ParsePKCS8PrivateKey(der)
	if err != nil {
		err = errors.New("x509.ParsePKCS8PrivateKey error " + err.Error())
		return
	}
	aKeyBytes, err := rsa.DecryptPKCS1v15(rand.Reader, privateKey.(*rsa.PrivateKey), aesKeyCiphertext)
	if err != nil {
		err = errors.New("rsa.DecryptPKCS1v15 error " + err.Error())
		return
	}
	aKey := string(aKeyBytes)

	// SHA1验签
	sign := fmt.Sprintf("%x", sha1.Sum([]byte(s1+aKey)))
	if sign != s2 {
		err = errors.New("验签失败")
		return
	}

	// AES 解密
	cipherText, err := base64.StdEncoding.DecodeString(s1)
	if err != nil {
		err = errors.New("s1 Base64解码错误 " + err.Error())
		return
	}
	block, err := aes.NewCipher(aKeyBytes)
	if err != nil {
		err = errors.New("aes.NewCipher error " + err.Error())
		return
	}
	blockMode := cipher.NewCBCDecrypter(block, []byte(aKey[0:block.BlockSize()]))
	plainText := make([]byte, len(cipherText))
	blockMode.CryptBlocks(plainText, cipherText)
	plainText = UnPadding(plainText)

	// 明文结果
	plaintext = string(plainText)
	return
}

func Padding(plainText []byte, blockSize int) []byte {
	n := blockSize - len(plainText)%blockSize //计算要填充的长度
	temp := bytes.Repeat([]byte{byte(n)}, n)  //对原来的明文填充n个n
	plainText = append(plainText, temp...)
	return plainText
}

func UnPadding(cipherText []byte) []byte {
	end := cipherText[len(cipherText)-1]               //取出密文最后一个字节end
	cipherText = cipherText[:len(cipherText)-int(end)] //删除填充
	return cipherText
}

// Rand32 生成32位随机字符串 使用crypto/rand 随机赋值byte数组， 然后md5返回32位十六进制字符串
func rand32() string {
	var b = make([]byte, 48)
	if _, err := rand.Read(b); err != nil {
		return ""
	}
	return fmt.Sprintf("%x", md5.Sum(b))
}
